crates/proof-of-sql/src/base/database/join_util.rs

```diff
@@ -34,6 +46,145 @@ pub fn cross_join<'a, S: Scalar>(
+    {
+        return Err(TableOperationError::DuplicateColumn);
+    }
+    // Check that the number of rows is good
+    for column in left_on {
+        assert_eq!(column.len(), left_num_rows);
+    }
+    for column in right_on {
+        assert_eq!(column.len(), right_num_rows);
+    }
+    // First of all sort the tables by the columns we are joining on
+    let left_indexes =
+        (0..left.num_rows()).sorted_unstable_by(|&a, &b| compare_indexes_by_columns(left_on, a, b));
+    let right_indexes = (0..right.num_rows())
+        .sorted_unstable_by(|&a, &b| compare_indexes_by_columns(right_on, a, b));
+    // Collect the indexes of the rows that match
+    let mut left_iter = left_indexes.into_iter().peekable();
+    let mut right_iter = right_indexes.into_iter().peekable();
+    let index_pairs = core::iter::from_fn(|| {
+        // If we have reached the end of either table, we are done
+        let (&left_index, &right_index) = (left_iter.peek()?, right_iter.peek()?);
+        match compare_single_row_of_tables(left_on, right_on, left_index, right_index).ok()? {
+            Ordering::Less => {
+                // Move left forward, return no pairs for this iteration
+                left_iter.next();
+                Some(Vec::new())
+            }
+            Ordering::Greater => {
+                // Move right forward, return no pairs for this iteration
+                right_iter.next();
+                Some(Vec::new())
+            }
+            Ordering::Equal => {
+                // Identify groups of equal keys from both sides
+                let left_group = left_iter
+                    .clone()
+                    .take_while(|&lidx| {
+                        compare_indexes_by_columns(left_on, left_index, lidx) == Ordering::Equal
+                    })
+                    .collect::<Vec<_>>();
+
+                let right_group = right_iter
+                    .clone()
+                    .take_while(|&ridx| {
+                        compare_indexes_by_columns(right_on, right_index, ridx) == Ordering::Equal
+                    })
+                    .collect::<Vec<_>>();
+
+                // All combinations of left_group x right_group
+                let pairs = left_group
+                    .iter()
+                    .cartesian_product(right_group.iter())
+                    .map(|(&l, &r)| (l, r))
+                    .collect::<Vec<_>>();
+
+                // Advance the iterators past the groups
+                left_iter.nth(left_group.len() - 1);
+                right_iter.nth(right_group.len() - 1);
+
+                Some(pairs)
+            }
+        }
+    })
+    // Flatten out the Vec<Vec<...>> from above into a single Vec
+    .flatten()
+    .collect::<Vec<_>>();
+    // Now we have the indexes of the rows that match, we can create the new table
+    let (left_indexes, right_indexes): (Vec<usize>, Vec<usize>) = index_pairs.into_iter().unzip();
```

Comment:
I would like to see this extracted into a function with the signature
`(left_on: &[Column<'a, S>], right_on: &[Column<'a, S>]) -> impl Iterator<Item = (usize, usize)>`
The only thing that needs to be `clone`d, `copy`ed, or `collect`ed inside that function is the `left_indexes` and `right_indexes` so that it can be sorted. (And even that can be hidden by using https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.sorted.)